source/tutorials/integration-testing-using-virtual-machines.rst

```diff
@@ -0,0 +1,228 @@
+            grant select on ${schema}.${table} to ${webRole};
+
+            create role ${username} inherit login password '${password}';
+            grant ${webRole} to ${username};
+          '';
+        };
+
+        users = {
+          mutableUsers = false;
+          users = {
+            # For ease of debugging the VM as the `root` user
+            root.password = "";
+
+            # Create a system user that matches the database user so that we
+            # can use peer authentication.  The tutorial defines a password,
+            # but it's not necessary.
+            "${username}".isSystemUser = true;
+          };
+        };
+
+        systemd.services.postgrest = {
+          wantedBy = [ "multi-user.target" ];
+          after = [ "postgresql.service" ];
+          script =
+            let
+              configuration = pkgs.writeText "tutorial.conf" ''
+                  db-uri = "postgres://${username}:${password}@localhost:${toString config.services.postgresql.port}/${database}"
+                  db-schema = "${schema}"
+                  db-anon-role = "${username}"
+              '';
+            in "${pkgs.haskellPackages.postgrest}/bin/postgrest ${configuration}";
+          serviceConfig.User = username;
+        };
+      };
+
+      client = { 
+        imports = [ sharedModule ];
+      };
+    };
+
+    testScript = ''
+      import json
+      import sys
+
+      start_all()
+
+      server.wait_for_open_port(${toString postgrestPort})
+
+      expected = [
+          {"id": 1, "done": False, "task": "finish tutorial 0", "due": None},
+          {"id": 2, "done": False, "task": "pat self on back", "due": None},
+      ]
+
+      actual = json.loads(
+          client.succeed(
+              "${pkgs.curl}/bin/curl http://server:${toString postgrestPort}/${table}"
+          )
+      )
+
+      if expected != actual:
+          sys.exit(1)
+    '';
+  })
+
+A few notes:
+
+- Between the machines defined inside the ``nodes`` attribute, hostnames 
+  are resolved based on their attribute names. In this case we have ``client`` and ``server``.
+
+- The testing framework exposes a wide set of operations used inside the ``testScript``.
+  A full set of testing operations is part of  
+  `VM testing operations API Reference <https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests>`_.
+
+
+Running tests
+-------------
+
+To set up all machines and execute the test script:
+
+.. code:: shell-session 
+
+  $ nix-build postgrest.nix
+
+You'll notice an error message if something goes wrong.
+
+In case the tests succeed, you should see at the end:
+
+.. code:: shell-session
+
+  1 out of 1 tests succeeded
+
+
+Developing and debugging tests
+------------------------------
+
+When developing tests or when something breaks, it's useful to interactively fiddle
+with the script or access a terminal for a machine.
+
+To interactively start a Python session with testing framework:
```

Comment:
I try hard to not offer two solutions to the same problem and since `nixos-run-vms` doesn't allow fiddling with the test script, it's inferior to the driver method. Does that make sense?